crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -279,66 +299,385 @@ where
+            _ => return Ok(token.as_ref().to_string()),
+        };
+
+        if let Some(owner) = owner {
+            let ibc_trace_key = ibc_trace_key(owner.to_string(), &hash);
+            if let Some(ibc_denom) = self.ctx.read_pre(&ibc_trace_key)? {
+                return Ok(ibc_denom);
+            }
+        }
+
+        // No owner is specified or the owner doesn't have the token
+        let ibc_denom_prefix = ibc_trace_key_prefix(None);
+        let ibc_denoms = self.ctx.iter_prefix(&ibc_denom_prefix)?;
+        for (key, ibc_denom, _gas) in ibc_denoms {
+            if let Some((_, token_hash)) =
+                is_ibc_trace_key(&Key::parse(key).into_storage_result()?)
+            {
+                if token_hash == hash {
+                    return String::try_from_slice(&ibc_denom[..])
+                        .into_storage_result()
+                        .map_err(Error::NativeVpError);
+                }
+            }
+        }
+
+        Ok(token.as_ref().to_string())
+    }
+
+    // Find the given IBC message in the changed keys and return the associated
+    // sequence number
+    fn search_ibc_transfer(
+        &self,
+        message: &IbcMsgTransfer,
+        keys_changed: &BTreeSet<Key>,
+    ) -> Result<Option<Sequence>> {
+        // Compute the packet commitment for this message
+        let packet_data_bytes = serde_json::to_vec(&message.packet_data)
+            .map_err(native_vp::Error::new)?;
+        let packet_commitment =
+            VpValidationContext::<'a, 'a, S, CA>::compute_packet_commitment(
+                &packet_data_bytes,
+                &message.timeout_height_on_b,
+                &message.timeout_timestamp_on_b,
+            );
+        // Try to find a key change with the same port, channel, and commitment
+        // as this message and note its sequence number
+        for key in keys_changed {
+            let Some(path) = is_ibc_commitment_key(key) else {
+                continue;
+            };
+            if path.port_id == message.port_id_on_a
+                && path.channel_id == message.chan_id_on_a
+            {
+                let Some(storage_commitment): Option<PacketCommitment> =
+                    self.ctx.read_bytes_post(key)?.map(Into::into)
+                else {
+                    // Ignore this event if it does not exist
+                    continue;
+                };
+                if packet_commitment == storage_commitment {
+                    return Ok(Some(path.sequence));
+                }
+            }
+        }
+        Ok(None)
+    }
+
+    // Try to determine which address would cause query_ibc_denom to yield the
+    // supplied denom
+    fn reverse_query_ibc_denom(
+        denom: &PrefixedDenom,
+        ibc_denoms: &BTreeMap<String, Address>,
+    ) -> Option<Address> {
+        ibc_denoms
+            .get(&denom.to_string())
+            .cloned()
+            // If the reverse lookup failed, then guess the Address
```

Comment:
Should we check that `query_ibc_denom` does in fact return the supplied denom? Or is that guaranteed?

Reply:
I think this should be guaranteed given that we only ever insert the results of `query_ibc_denom` into the `ibc_denoms` map. My intent had been that nothing other than the outputs of `query_ibc_denom` is ever inserted into the `ibc_denoms` map.

Reply:
This map has now been removed since the suggestion https://github.com/anoma/namada/pull/3409#discussion_r1644247178 is now being used.